### PR TITLE
Transaction max resends control

### PR DESCRIPTION
**Pull request by paulobrizolara**
This commit adds some fields to a transaction that enables control
the max number of messages that will be sent after the transaction
receives a response.

This is useful to allow multicast transactions (that keep sending
requests after receives a response) to stop (or limit) resending,
but keep waiting for other responses.

### CODE REVIEW

**Comment by paulobrizolara** [comment]
@darconeous, i am doing this enhancement proposal to enable users limit the amount of resendings done in a multicast request (after receive a response).
When you can, take a look and tell me what you think. If you agree and if so, what needs to be changed.
Thanks!

File: src/examples/example-multicast-request.c [hunk 1]
```diff
@@ -134,6 +134,9 @@ main(int argc, char ** argv)
 		exit(EXIT_FAILURE);
 	}
 
+	//Send no more than one message (or set of bursts) after receives a response
+	transaction->max_resends = 1;
```
Comment (paulobrizolara) [inline src/examples/example-multicast-request.c]: I have changed the example to shows the use of the feature.

File: src/smcp/smcp-transaction.c [hunk 2]
```diff
@@ -219,6 +219,81 @@ smcp_transaction_new_msg_id(
...
+
+		//Send full burst
+		|| (should_burst
+				&& resends == handler->max_resends
+				&& (handler->attemptCount % SMCP_TRANSACTION_BURST_COUNT != 0) //middle of burst
+				);
+}
+
+static
+smcp_status_t smcp_internal_resend(smcp_t self, smcp_transaction_t handler, smcp_cms_t * cms, void* context)
```
Comment (paulobrizolara) [inline src/smcp/smcp-transaction.c]: I have extracted the code that does the resend of the transaction into a function to enhance the legibility.

**Comment by darconeous** [comment]
Ooo, nice!

I'll take a closer look over the next day or two, but overall this looks like a good change. Having the ability to limit retransmits per transaction seems like a useful capability.

The refactoring/cleanup is nice, too. It improves readability substantially.

**Comment by darconeous** [comment]
There is a lot I like about this PR. I especially like the code reorganization. That being said, I feel like there are some additional improvements we can make:

The new behavior you are defining is a max-resend-after-received-reply, but there is no guarantee if getting any replies at all from multicast requests. I don't really think whether you have received a reply or not is a useful indication to determine if you should change your retransmit strategy to start counting down retransmits. I think it makes more sense to have an overall maximum retransmit count: a way to specify a value other than `SMCP_TRANSACTION_MAX_ATTEMPTS` for specific transactions.

I'd like to avoid having to directly reach into the transaction structure to use this feature. In most cases the contents should be considered private. I'm not wild about adding another function, but that might be the way to go.

Thoughts? Again, thanks for contributing!

**Comment by paulobrizolara** [comment]
An overall modifiable maximum retransmit may be a more simple approach to solve the case.
However, i think it may not give enough control to clients.

I did this way to be somewhat similar to how unicast messages are handled.
They keep sending messages until receive a response. 
Then a multicast request would send more few times (controlled by max resends) to ensure all nodes received the response. 

A strategy that I had not thought of before is just let the 'resend callback' to control this behaviour.
Looks like this function can just not send the package. Right? (not tested yet)
So, the library could expose some info about the transaction (ex.: if a response was received or the number of received responses, and maybe other things).
This could be made through some functions or splitting the transaction into a public and a private part.

So, what approach do you prefer?
I can change this code or make another PR.

**Comment by darconeous** [comment]
> I did this way to be somewhat similar to how unicast messages are handled.
They keep sending messages until receive a response. Then a multicast request would send more few times (controlled by max resends) to ensure all nodes received the response.

It is still not clear to me why receiving a response should have any effect on multicast retransmits. Why would that behavior be desirable? What sorts of cases would it be helpful for that wouldn't be similarly addressed by changing the max retransmits for the transaction?

It just doesn't seem like it would be more beneficial than the general case of a configurable maximum number of retransmits. I actually think it is detrimental, because multicast requests don't have to generate responses—in which case there would be no way to configure the maximum number of retransmits per transaction.

> A strategy that I had not thought of before is just let the 'resend callback' to control this behaviour. Looks like this function can just not send the package. Right? (not tested yet)

Correct. The user could implement this behavior in the callbacks. I like this approach better, because it doesn't encourage weird behavior but doesn't prevent you from doing it for experimental purposes.

> This could be made through some functions or splitting the transaction struct into a public and a private part (and giving to the callback an access to it).

Meh, in this case I think it is such a weird behavior I would be fine with someone having to dig into the structure to implement it on the client side. ABI stability isn't going to be a major concern until 1.00.00 anyway, so it isn't a big deal.

**Comment by paulobrizolara** [comment]
Ok.
Since this behaviour can be implemented most on client side (using the resend callback), i will remove the 'max-resends' behaviour from PR and set an overall maximum retransmit.

Also, i am thinking in include another status value to allow the resend callback to explicitly communicate: 'stop resending but continue to wait responses'. After return this status, the resend callback would not be called, but the transaction would not be finished.
Ex.:

            smcp_status_t resend_callback(void * context){
                   //...
                   if(!continueResend()){
                          return SMCP_STATUS_STOP_RESENDING;
                   }
                   //...
            }

What do you think?

**Comment by darconeous** [comment]
I think `SMCP_STATUS_STOP_RESENDING` is a great idea! 😊

**Comment by paulobrizolara** [comment]
Nice! I will add in a while.

**Comment by paulobrizolara** [comment]
Done!
I had also changed the example a bit to show the feature.

**Comment by darconeous** [comment]
I think this commit looks good. There are some minor style nits, but then again there are style nits all over SMCP (I need to write up a style guide at some point). I can fix those easy enough in a follow up commit, so I'm not going to cause that to hold this back.

👍

**Pull request merged**